### PR TITLE
*: query failed after add index / timestamp out-of-range (#28424)

### LINKED ISSUE #28424: query failed after add index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
set time_zone='UTC';
drop table if exists t02;
create table t02(a timestamp);
select * from t02 where a != "2038-1-19 3:14:45";
alter table t02 add unique index (a);
select * from t02 where a != "2038-1-19 3:14:45";
```

### 2. What did you expect to see? (Required)
query successfully after adding index.
### 3. What did you see instead (Required)
```sql
ySQL root@127.0.0.1:test> set time_zone='UTC';
                       ->  drop table if exists t02;
                       ->  create table t02(a timestamp);
                       ->  select * from t02 where a != "2038-1-19 3:14:45";
                       ->  alter table t02 add unique index (a);
                       ->  select * from t02 where a != "2038-1-19 3:14:45";
Query OK, 0 rows affected
Time: 0.002s

Query OK, 0 rows affected
Time: 0.012s

Query OK, 0 rows affected
Time: 0.006s

+---+
| a |
+---+

0 rows in set
Time: 0.014s

Query OK, 0 rows affected
Time: 2.546s
(1292, "Incorrect timestamp value: '2038-01-19 03:14:45.000000'")
MySQL root@127.0.0.1:test>

```
### 4. What is your TiDB version? (Required)
master (76267b21aee467f334)
v5.2.1

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #28424 

Problem Summary: Timestamp out-of-range and range optimizer incompatibility.

### What is changed and how it works?

Two issues, convertToMysqlTimestamp did set out-of-range timestamp to
Zero, when it would still be a storable/comparable KindMysqlTime.
(we only artificially limit the range of Timestamp to be compatible
with MySQL).

When this change was done,
it broke #26584 (insert of invalid timestamp succeeded),
which needed a different change in handleZeroDatetime

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fixed index handling bug with timestamp out-of-range.
```

### CODE REVIEW

File: util/ranger/ranger.go [hunk 1]
```diff
@@ -102,6 +102,9 @@ func convertPoint(sc *stmtctx.StatementContext, point *point, tp *types.FieldTyp
 			// Ignore the types.ErrOverflow when we convert TypeNewDecimal values.
 			// A trimmed valid boundary point value would be returned then. Accordingly, the `excl` of the point
 			// would be adjusted. Impossible ranges would be skipped by the `validInterval` call later.
+		} else if point.value.Kind() == types.KindMysqlTime && tp.Tp == mysql.TypeTimestamp && terror.ErrorEqual(err, types.ErrWrongValue) {
```
Comment (tiancaiamao) [inline util/ranger/ranger.go:105]: Is this the only case that return `types.ErrWrongValue` and can be saftly ignored?
If in some other cases `ErrWrongValue` is also returned, and we ignore the error unexpectedly

**Review by tiancaiamao — APPROVED** [review]

File: util/ranger/ranger.go (hunk 1 above)
Comment (mjonss) [reply util/ranger/ranger.go:105]: I think so.

From what I can see, the `ConvertTo(sc, tp)` with `Kind() == types.KindMysqlTime` and `tp.Tp == mysql.TypeTimestamp` can only end up in the part of types/datum.go above, that I also changed in this PR, so I cannot see any other case that would end up being ignored here.

I.e. it is only points of MysqlTime converted to Timestamp that would generate an error in `GetMysqlTime().Convert()` that would be matched here.

**Comment by mjonss** [comment]
@AilinKid or @tangenta could one of you please take a look?

**Review by xhebox — APPROVED** [review]

**Comment by guo-shaoge** [comment]
> /label needs-cherry-pick-5.2 /label needs-cherry-pick-5.3

Looks like it's a corner case bug. Maybe we can close other cherry pick pr.

**Pull request merged**